arrow-docs/docs/fx/managing-state/README.md

```diff
@@ -0,0 +1,188 @@
+---
+layout: docs-fx
+title: Arrow Fx - Handling concurrent state
+permalink: /fx/managing-state/
+---
+
+# Handling concurrent state
+
+Fx programs have a variety of options available when it comes to managing state in a concurrent environment.
+
+## Problems with shared state in a concurrent environment
+
+The problem of having shared state throughout an application is consistency when this state is read and updated.
+An invariant checked a few lines ago may now no longer be valid if other threads changed the variables that were read.
+Updating a variable based on a prior read may lead to inconsistent state if there has been another update in-between...
+
+These issues are referred to as race-conditions and are the source of many headaches and lots of time spent debugging.
+
+In this article we will explore how arrow fx helps to deal with this issue and in some cases can even avoid it entirely.
+
+## Accessing a single shared variable
+
+When we are accessing only a single variable we can use [Atomic](TODO Link).
+This gives us access to a single mutable variable where reads and writes are atomic operations,
+ meaning no other access to the variable can happen during any operation on an atomic variable.
+ 
+```kotlin:ank:playground
+import arrow.fx.coroutines.*
+
+suspend fun main() {
+  val count = Atomic(0)
+  (0 until 20_000).parTraverse {
+    count.update(Int::inc)
+  }
+  println(count.get())
+}
+```
+
+In this example `count.update` is our atomic operation, and it is performed by possibly many threads at the same time.
+Despite that our counter increments one by one and never contains an invalid state (like say two threads incrementing 1 to 2 at the same time and both persisting 2 missing one update).
+
+### Drawbacks of using atomic variables
+
+Atomic variables have one major drawback: Their atomic guarantees do not scale beyond one variable.
+We can guarantee that access to one variable is atomic, but accessing two variables with atomic operations, even right after each other, will not make the entire operation atomic.
+
+## Securing access to shared state with locks
```

Comment:
This can also be done with `val mutex = Semaphore(1)`, but that's a different pattern than `ConcurrentVar`.

Reply:
`Semaphore(1) == ConcurrentVar(Unit)` right? Either way it has the same drawbacks

Reply:
`Semaphore(1) == ConcurrentVar(Unit)` 🤔 I have never thought about it like this 😅

I guess in this way `acquire == take` and `release` == `put`where they'er all scheduled in order. Anyway, you almost always want to use `withPermit` which is build upon bracketing `acquire` with `release` to make sure you don't lose any permits. Which is like ensuring you always correctly release your locks. (Not to be confused with JVM-level locks, which aren't used in either `ConcurrentVar` nor `Semaphore`).

Reply:
Yeah `withPermit` is definitely the best pattern (haskell recommends `withMVar` over anything else as well^^). Do we also have a `withConcurrentVar` method? Could use the same treatment.

Reply:
No, we currently don't have such a method but that would be a great addition I think!